lib/httplib/csrf/csrf.go

```diff
@@ -0,0 +1,138 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csrf
+
+import (
+	"crypto/subtle"
+	"encoding/hex"
+	"net/http"
+
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+)
+
+const (
+	// CookieName is a name of the cookie
+	CookieName = "grv_csrf"
+	// HeaderName is the default HTTP request header to inspect
+	HeaderName = "X-CSRF-Token"
+	// tokenLenBytes is CSRF token length in bytes.
+	tokenLenBytes = 32
+	// defaultMaxAge is the default MaxAge for cookies.
+	defaultMaxAge = 0
+)
+
+// AddCSRFProtection adds CSRF token into the user session via secure cookie,
+// it implements "double submit cookie" approach to check against csrf attacks
+// https://www.owasp.org/index.php/Cross-Site_Request_Forgery_%28CSRF%29_Prevention_Cheat_Sheet#Double_Submit_Cookie
+func AddCSRFProtection(w http.ResponseWriter, r *http.Request) (string, error) {
+	encodedToken := ""
+	token, err := extractFromCookie(r)
+	// if there was an error retrieving the token, the token doesn't exist
+	if err != nil || len(token) != tokenLenBytes {
+		encodedToken, err = utils.CryptoRandomHex(tokenLenBytes)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+	} else {
+		encodedToken = hex.EncodeToString(token)
+	}
+
+	save(encodedToken, w)
+	return encodedToken, nil
+}
+
+// VerifyToken checks if the cookie value and request value match.
+func VerifyToken(w http.ResponseWriter, r *http.Request) error {
+	realToken, err := extractFromCookie(r)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if len(realToken) != tokenLenBytes {
+		return trace.BadParameter("invalid CSRF token length")
+	}
+
+	// compare the request token against the real token
+	requestToken := requestToken(r)
+	if !compareTokens(requestToken, realToken) {
+		return trace.BadParameter("CSRF tokens do not match")
+	}
+
+	return nil
+}
+
+// extractFromCookie retrieves a CSRF token from the session cookie.
+func extractFromCookie(r *http.Request) ([]byte, error) {
+	// Retrieve the cookie from the request
+	cookie, err := r.Cookie(CookieName)
+	if err != nil {
+		return nil, trace.BadParameter("cannot find CSRF cookie", err)
+	}
+
+	token, err := decode(cookie.Value)
+	if err != nil {
+		return nil, trace.BadParameter("cannot decode CSRF cookie value", err)
+	}
+
+	return token, nil
+}
+
+// save stores encoded CSRF token in the session cookie.
+func save(encodedToken string, w http.ResponseWriter) string {
+	cookie := &http.Cookie{
+		Name:     CookieName,
+		Value:    encodedToken,
+		MaxAge:   defaultMaxAge,
+		HttpOnly: true,
+		Secure:   true,
+		Path:     "/",
+	}
+
+	// write the authenticated cookie to the response.
+	http.SetCookie(w, cookie)
+	w.Header().Add("Vary", "Cookie")
+	return encodedToken
+}
+
+// requestToken returns the issued token from HTTP header.
+func requestToken(r *http.Request) []byte {
+	issued := r.Header.Get(HeaderName)
+	decoded, err := decode(issued)
+	if err != nil {
```

Comment:
can you rename `requestToken` to `extractToken` and return proper error, letting higher level function to process it.